examples/next-sse-chat/src/server/routers/channel.ts

```diff
@@ -0,0 +1,133 @@
+  once<TEv extends keyof MyEvents>(event: TEv, listener: MyEvents[TEv]): this;
+  emit<TEv extends keyof MyEvents>(
+    event: TEv,
+    ...args: Parameters<MyEvents[TEv]>
+  ): boolean;
+}
+
+class MyEventEmitter extends EventEmitter {
+  public toIterable<TEv extends keyof MyEvents>(
+    event: TEv,
+  ): AsyncIterable<Parameters<MyEvents[TEv]>> {
+    return on(this, event);
+  }
+}
+
+// In a real app, you'd probably use Redis or something
+export const ee = new MyEventEmitter();
+
+// who is currently typing for each channel, key is `name`
+export const currentlyTyping: Record<string, WhoIsTyping> = Object.create(null);
+
+// every 1s, clear old "isTyping"
+setInterval(() => {
+  const updatedChannels = new Set<string>();
+  const now = Date.now();
+  for (const [channelId, typers] of Object.entries(currentlyTyping)) {
+    for (const [key, value] of Object.entries(typers ?? {})) {
+      if (now - value.lastTyped.getTime() > 3e3) {
+        delete typers[key];
+        updatedChannels.add(channelId);
+      }
+    }
+  }
+  updatedChannels.forEach((channelId) => {
+    ee.emit('isTypingUpdate', channelId, currentlyTyping[channelId] ?? {});
+  });
+}, 3e3).unref();
+
+export const channelRouter = {
+  list: publicProcedure.query(() => {
+    return db.query.Channel.findMany();
+  }),
+
+  create: authedProcedure
+    .input(z.object({ name: z.string() }))
+    .mutation(async ({ ctx, input }) => {
+      const [channel] = await db
+        .insert(Channel)
+        .values({
+          name: input.name,
+        })
+        .returning();
+
+      return channel.id;
+    }),
+
+  isTyping: authedProcedure
+    .input(z.object({ channelId: z.string().uuid(), typing: z.boolean() }))
+    .mutation(async (opts) => {
+      const { name } = opts.ctx.user;
+      const { channelId } = opts.input;
+
+      if (!currentlyTyping[channelId]) {
+        currentlyTyping[channelId] = {};
+      }
+
+      if (!opts.input.typing) {
+        delete currentlyTyping[channelId][name];
+      } else {
+        currentlyTyping[channelId][name] = {
+          lastTyped: new Date(),
+        };
+      }
+      ee.emit('isTypingUpdate', channelId, currentlyTyping[channelId]);
+    }),
+
+  whoIsTyping: publicProcedure
+    .input(
+      z.object({
+        channelId: z.string().uuid(),
+        lastEventId: z.string().optional(),
+      }),
+    )
+    .subscription(async function* (opts) {
+      const { channelId } = opts.input;
+      let lastEventId = opts?.input?.lastEventId ?? '';
+
+      if (!currentlyTyping[channelId]) {
+        currentlyTyping[channelId] = {};
+      }
+
+      const maybeYield = function* (who: WhoIsTyping) {
+        const id = Object.keys(who).sort().toString();
+        if (lastEventId === id) {
+          return;
+        }
+        yield {
+          id,
+          data: Object.keys(who).filter(
+            (user) => user !== opts.ctx.session?.user?.name,
+          ),
+        } satisfies SSEvent;
```

Comment:
What would you think about letting you `yield` anything and have a special function when you wanna yield `SSEvent`?

e.g.

```ts
yield { foo: "bar" }
```

then `data` would simply be `"bar"`

and doing eg

```ts
yield sseEvent({
  id: 1,
  foo: "bar",
})
```
if you wanna `yield` a "proper" SSE event

Reply:
I like that though instead of sseEvent, how about just sse or ssEvent or serverSentEvent?

Reply:
sure, not bothered about the naming